util/src/RawMemoryPool.cpp

```diff
@@ -0,0 +1,142 @@
+  maxChunksNum_ = maxChunksNum;
+  pool_ = make_shared<boost::lockfree::queue<char*, boost::lockfree::fixed_sized<true>>>(maxChunksNum_);
+  for (int32_t i = 0; i < initialChunksNum_; ++i) {
+    pool_->push(new char[chunkSize_]);
+    metrics_.increaseAvailableChunksNum();
+    metrics_.increaseAllocatedChunksNum();
+  }
+  numOfAvailableChunks_ = initialChunksNum;
+  numOfAllocatedChunks_ = initialChunksNum;
+  LOG_INFO(logger(),
+           "Memory pool allocated" << KVLOG(chunkSize_,
+                                            initialChunksNum_,
+                                            maxChunksNum_,
+                                            numOfAllocatedChunks_,
+                                            numOfAvailableChunks_,
+                                            maxAvailableChunksPercentage_));
+}
+
+void RawMemoryPool::increaseNumOfAllocatedChunks() {
+  numOfAllocatedChunks_++;
+  metrics_.increaseAllocatedChunksNum();
+}
+
+void RawMemoryPool::increaseNumOfAvailableChunks() {
+  numOfAvailableChunks_++;
+  metrics_.increaseAvailableChunksNum();
+}
+
+void RawMemoryPool::decreaseNumOfAllocatedChunks() {
+  numOfAllocatedChunks_--;
+  metrics_.decreaseAllocatedChunksNum();
+}
+
+void RawMemoryPool::decreaseNumOfAvailableChunks() {
+  numOfAvailableChunks_--;
+  metrics_.decreaseAvailableChunksNum();
+}
+
+char* RawMemoryPool::allocateChunk() {
+  char* chunk = nullptr;
+  if (numOfAllocatedChunks_ < maxChunksNum_) {
+    chunk = new char[chunkSize_];
+    increaseNumOfAllocatedChunks();
+    increaseNumOfAvailableChunks();
+  }
+  if (chunk) LOG_DEBUG(logger(), "A chunk has been allocated" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+  return chunk;
+}
+
+void RawMemoryPool::deleteChunk(char*& chunk) {
+  delete[] chunk;
+  decreaseNumOfAllocatedChunks();
+  LOG_DEBUG(logger(), "A chunk has been deleted" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+}
+
+void RawMemoryPool::returnChunkToThePool(char* chunk) {
+  pool_->push(chunk);
+  increaseNumOfAvailableChunks();
+  LOG_DEBUG(logger(), "A chunk has been returned to the pool" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+  waitForAvailChunkCond_.notify_one();
+}
+
+bool RawMemoryPool::isPoolPruningRequired() {
+  return (((numOfAllocatedChunks_ > initialChunksNum_) &&
+           (numOfAvailableChunks_ >= numOfAllocatedChunks_ * maxAvailableChunksPercentage_)));
+}
+
+char* RawMemoryPool::getChunk() {
+  ConcordAssert(pool_.get() != nullptr);
+  char* chunk = nullptr;
+  if (!pool_->pop(chunk)) {
+    // No available chunks => allocate a new one, if permitted
+    chunk = allocateChunk();
+    if (!chunk) {
+      // Pool size limit has been reached; wait until some chunk gets released
+      unique_lock<mutex> lock(waitForAvailChunkLock_);
+      while (!stopWorking_ && !numOfAvailableChunks_)
+        waitForAvailChunkCond_.wait_until(lock, steady_clock::now() + milliseconds(WAIT_TIMEOUT_MILLI));
```

Comment:
shouldn't this thread need to pop again after the notification? 
maybe the while should be something like this
```
while (!stopWorking_ && !chunk){
        waitForAvailChunkCond_.wait_until(lock, steady_clock::now() + milliseconds(WAIT_TIMEOUT_MILLI));
         pool_->pop(chunk);
}
````
wdyt?

Reply:
Agree